File: nodejs/src/strategy/EWalletPayment.ts
import { PaymentStrategy, PaymentResult } from "./PaymentStrategy";

class EWalletPayment implements PaymentStrategy {
    private walletName: string;
    private balance: number;

    constructor(walletName: string, balance: number) {
        this.walletName = walletName;
        this.balance = balance;
    }

    getName(): string { return `Ví ${this.walletName}`; }

    pay(amount: number): PaymentResult {
        if (this.balance < amount) {
            return { success: false, message: "Số dư không đủ" };
        }
        this.balance -= amount;
        return {
            success: true,
            message: `Thanh toán ${amount.toLocaleString()} VND. Số dư còn: ${this.balance.toLocaleString()} VND`,
        };
    }
}

export default EWalletPayment;

File: spring/demo/src/main/java/com/example/demo/strategy/EWalletPayment.java
package com.example.demo.strategy;

public class EWalletPayment implements PaymentStrategy {
    private final String walletName;
    private int balance;

    public EWalletPayment(String walletName, int balance) {
        this.walletName = walletName;
        this.balance = balance;
    }

    @Override
    public String getName() {
        return "Ví " + walletName;
    }

    @Override
    public PaymentResult pay(int amount) {
        if (balance < amount) {
            return new PaymentResult(false, "Số dư không đủ");
        }

        balance -= amount;
        return new PaymentResult(true, String.format("Thanh toán %,d VND. Số dư còn: %,d VND", amount, balance));
    }
}
